Various updates and refactors to build platform in general

touhidurrr opened the pull request:
What this PR contains:

1. `Gradle` version `8.11.1` -> `9.0.0` (latest stable)
2. Update some actions versions in `dockerPublish.yml`
3. `Dockerfile` improvements

yairm210 commented:
The docker failures seem real
Wet may need to upgrade gradle in the docker image

touhidurrr commented:
> The docker failures seem real
> Wet may need to upgrade gradle in the docker image

It is failing in my local gradle 9.0.0 also. Replacing `tasks.create()` with `tasks.register()` might not be so simple. I also wanted to try `dependsOn()` but the comments seem to be advicing against it.

For now, I will simply revert `tasks.create()` -> `tasks.register()` change later, rename this as a minor refactor pr. And maybe learn how tasks work in gradle and open another pr later.

touhidurrr commented:
> > when did you guys revert Kotlin version.
> 
> [As far as I can see, it's still 2.1.20](https://github.com/yairm210/Unciv/blob/master/buildSrc/src/main/kotlin/BuildConfig.kt#L5), unchanged since #13602

I am confused, then what did I change in 6e9a8746f104208855599e8e9b4407b13d18a1da

touhidurrr commented:
ok @yairm210, now I think we need to update the gradle version inside docker image.

touhidurrr commented:
Really confused where these errors come from:
<img width="889" height="176" alt="image" src="https://github.com/user-attachments/assets/d2a7f2b3-40e0-4c20-bbec-dd486db21229" />

yairm210 commented:
Too much going on here at once, which will make it really hard to figure out what broke if something breaks

I see several different things going on, each of which is basically a different subject and therefore PR

- CI upgrades / minor changes - version updates are harmless, if->when is harmless, if it breaks will not affect end users (will just break release process), approved
- Library updates - Approved but we need to keep an eye on potential breakages
- Gradle updates - More involved, unclear what the implications are
- Docker changes - most of these are basically nothing, unclear why you feel compelled to make these changes

yairm210 commented:
Also
- Did you test the docker actually works, and not just "is buildable"?
- Did you test that the various packr outputs are created?

touhidurrr commented:
> Minor things: Isn't Enum.values() deprecated by now? And wouldn't the jre/jdk file name be more elegant as enum property than via when?

Ok so I was able to figure out why did it happen. Basically, gradle scripts run on their own kotlin version separate from the projects kotlin version. So, even if the IDE was showing a deprecation warning, it was not correct. I guess the IDE only loads one Kotlin version at a time and that is the projects version.

So, in short the way to make `Enum.entries` work is to increase the gradle version which in turn will increase the kotlin version that gradle uses. `9.0.0` just happens to use the same kotlin version series (`2.1.x`) that we currently use so that fixes it.

touhidurrr commented:
> Docker changes - most of these are basically nothing, unclear why you feel compelled to make these changes

I touched it because some previous builds were only failing on Docker but not anywhere else. Then I realized that nothing there was fixing the issue of hardcoded kotlin version not working in docker. So, I just did some refractors that I saw possible and left it there. A list of such changes would be:

1. Remove unnecessary gradle cannot detect jdk patch. The issue mentioned there was close long ago and removing those lines did not affect gradle in any way.
2. Use adoptium api to always fetch the latest version of JRE 11 instead of an old hardcoded version. Again harmless but something that I spotted and fixed.
3. Fixed some docker complaints. Such as `AS` has to be the same case  as `FROM`.
4. Broken `&&` chain into separate `RUN` calls. It just makes debugging a little easier.

And yes, I did not test if it actually runs. I will try when I go back home and give an update.

touhidurrr commented:
ok @yairm210, I tried and it simply does not work on windows. So, cannot tell if it is working.

![](https://cdn.discordapp.com/attachments/654043902948409394/1404139220070830131/image.png?ex=689a1a0d&is=6898c88d&hm=567da6479b8161343e3165bc9943cf1050cea04563b8c419758cb419857d76d9)

yairm210 commented:
Did you connect via vnc?
https://github.com/yairm210/Unciv?tab=readme-ov-file#run-with-docker-

touhidurrr commented:
> Did you connect via vnc? https://github.com/yairm210/Unciv?tab=readme-ov-file#run-with-docker-

Ok it worked. But very weird way to play Unciv for sure. I was expecting a window to pop up.

touhidurrr commented:
But anyways, building the image locally in Windows is still an impossibility. Thus I am unable to test if it works.

SeventhM commented:
> But anyways, building the image locally in Windows is still an impossibility

I have no clue if this is a factor and it probably shouldn't be, but I should ask...

What version of Android Studio do you have?

touhidurrr commented:
> though it's now suggesting to use Kotlin 2.2.0

Weird. From patch notes I am pretty sure Gradle 9 itself use Kotlin 2.1.x. So I wonder why are they suggesting 2.2. It did not suggest this for me.

touhidurrr commented:
> What version of Android Studio do you have?

Latest feature drop. I uodated it a few days ago. Also, this is likely because of how Docker Desktop works on Windows Home. It runs inside WSL. My guess is it cannot access Android Sdk binaries from WSL.

SeventhM commented:
> > Builds fine, no issues
> 
> Can you kindky verify if the docker image build and runs fine with exactly this PR's content? You do not need to disable purity.

Tbh, I have zero clue how to do anything with docker at all

touhidurrr commented:
> Tbh, I have zero clue how to do anything with docker at all

Me before opening this PR: Everyone use Docker. I am the only guy who doesn’t.
Me after this PR: What is Docker and Why is the maintainers even want me to test if stuff runs on it. Who uses such niche technology that crashes everywhere.

touhidurrr commented:
> Yes, please test that the docker works

Can you test this? As previously mentioned I am cannot build this in my system for some reason. Docker cannot use my Android Studio for some reason and complaining that it needs it.

yairm210 commented:
I don't think these changes are necessary in the first place, everything currently works fine
